### PR TITLE
try fitting with only Alter

### PR DESCRIPTION
**Pull request by murphymadeleine21**
This is close but it's having some issues with inflating the predictions back to original size. Not sure why this would be happening, for example with the ADCC case, dropped is size 84 and Y_pred is size 97, so `np.insert` should be working fine, but it's saying there's an out of bounds error. A bit confused, I've been looking at it for a while so maybe I'm missing something small.

### CODE REVIEW

**Comment by aarmey** [comment]
@scottdtaylor95 is going to start helping with improving the regression, and eventually look at other similar datasets. Perhaps he can help figure out the problem here.

**Comment by scottdtaylor95** [comment]
So I looked through it, and the first issue I noticed is that there's a discrepancy between X and Y in regression.py due to using X[np.where(Y)]. For certain functions (eg. ADCD) some values of Y are 0, so these get removed. Assuming you want to keep the locations where Y = 0, one fix would be deleting the X = X[indices, :] line and replacing np.where(Y) with Y.index. The second issue I found was in the evaluation part - I believe dropped should be passed as dropped and not dropped[0] to insert all dropped indices. Then in accuracy_alterSubj include the line Y = Y.to_numpy() right under indices = AlterIndices(), and then change the inflation to size to something like 
for i, drop_idx in enumerate(dropped):
     Ypred = np.insert(Ypred, drop_idx, np.nan)
     Y = np.insert(Y, drop_idx, np.nan)
This will avoid any out of bounds issues with trying to insert all at once. - There's probably also a way to do this without the loop by changing the dropped indices to match the Ypred/Y indices, if desired.
However the np.nan values will create problems later if you want to use the non Alter subjects, so maybe find a way to insert the dropped values as well (perhaps make dropped a dictionary of dropped indices and values?). I think that should make most of it work, at least up until that point.

**Comment by murphymadeleine21** [comment]
I think you are right that we need to have a dict of some sort that stores the original value too. That way, when we want to only analyze subjects NOT included in Alter, we won't have only Nans aka nothing to evaluate.

**Comment by murphymadeleine21** [comment]
Also, I'm coming to realize a bigger issue.. We can't compare predictions for "Not Alter" because Non-alter subjects never get predicted. So if we refill Y&Ypred where we didn't predict, the refilled values will always be the same (either Y original values or Nans), and thus there is no point in finding an accuracy score.

**Comment by aarmey** [comment]
The Alter subjects give you the model by fit(), then you can run the non-alter subjects through the model with predict(). The issue is you just can't set the model parameters using the non-alter subjects.

File: syserol/regression.py [hunk 1]
```diff
@@ -36,52 +36,40 @@ def function_elastic_net(function="ADCC"):
...
+    X_Alter = X[Y_Alter.index]
     Y = Y[np.isfinite(Y)]
+    X = X[Y.index]
+
+    enet = elasticNetFunc(X_Alter, Y_Alter)
 
-    Y_pred = elasticNetFunc(X, Y)
 
     if evaluation == "all":
+        Y_pred = cross_val_predict(enet, X, Y, cv=len(Y), n_jobs=-1)
```
Comment (aarmey) [inline syserol/regression.py]: You only have to use cross validation if the observations you’re predicting are in the training data. Non-Alter subjects aren’t, so you can just call predict()

**Review by aarmey — CHANGES_REQUESTED** [review]

File: syserol/regression.py (hunk 1 above)
Comment (aarmey) [reply syserol/regression.py]: Yes, this means you’ll have to separately handle the Alter and non-Alter subjects when collecting their predictions.

**Comment by murphymadeleine21** [comment]
This seems to get the sizing and prediction methods correct, but at least for the all case we're getting a Nan accuracy score (I think because of a negative r2 score)

**Comment by aarmey** [comment]
That doesn't make sense—there's no reason the prediction would be much worse. Are you **sure** it's because of negative r2, and not nan r2?

**Comment by murphymadeleine21** [comment]
Yes, when I print it out they are all real values. One note is that the `Y_pred_Alter` and `Y_pred_notAlter` are all larger values (around 10-15). Might be predicting differently without the crossval?

File: syserol/regression.py [hunk 2]
```diff
@@ -36,52 +36,43 @@ def function_elastic_net(function="ADCC"):
...
+    dropped = np.unique(np.concatenate((np.nonzero(np.isnan(Y.to_numpy()))[0], notAlter)))
+    Y_notAlter = Y[dropped][np.isfinite(Y)]
+    X_notAlter = X[Y_notAlter.index]
+    Y_Alter = Y[indices][np.isfinite(Y)]
+    X_Alter = X[Y_Alter.index]
     Y = Y[np.isfinite(Y)]
+    X = X[Y.index]
 
-    Y_pred = elasticNetFunc(X, Y)
+    enet = elasticNetFunc(X_Alter, Y_Alter)
```
Comment (aarmey) [inline syserol/regression.py]: Why use elasticNetFunc() anymore when you're only using a couple lines from it?

File: syserol/regression.py [hunk 3]
```diff
@@ -36,52 +36,43 @@ def function_elastic_net(function="ADCC"):
...
+            Y_pred_Alter = np.insert(Y_pred_Alter, drop_idx, Y_pred_notAlter[i])
+        Y_pred = Y_pred_Alter
         return Y, Y_pred, np.sqrt(r2_score(Y, Y_pred))
     elif evaluation == "Alter":
-        return accuracy_alterSubj(Y, Y_pred, dropped[0])
+        Y_pred = cross_val_predict(enet, X_Alter, Y_Alter, cv=len(Y_Alter), n_jobs=-1)
+        return Y_Alter, Y_pred, np.sqrt(r2_score(Y_Alter, Y_pred))
     elif evaluation == "notAlter":
-        return accuracy_alterSubj(Y, Y_pred, dropped[0], union=False)
+        Y_pred = enet.predict(X_notAlter)
```
Comment (aarmey) [inline syserol/regression.py]: Where are you fitting the model before you predict?

File: syserol/regression.py [hunk 4]
```diff
@@ -36,52 +36,43 @@ def function_elastic_net(function="ADCC"):
...
+    X_notAlter = X[Y_notAlter.index]
+    Y_Alter = Y[indices][np.isfinite(Y)]
+    X_Alter = X[Y_Alter.index]
     Y = Y[np.isfinite(Y)]
+    X = X[Y.index]
 
-    Y_pred = elasticNetFunc(X, Y)
+    enet = elasticNetFunc(X_Alter, Y_Alter)
 
     if evaluation == "all":
```
Comment (aarmey) [inline syserol/regression.py:54]: Why include "all" if we never use it?

File: syserol/regression.py [hunk 5]
```diff
@@ -36,52 +36,43 @@ def function_elastic_net(function="ADCC"):
     df_variables = df_merged.drop(["subject"] + functions, axis=1)
 
     # perform regression
-    Y_pred = elasticNetFunc(df_variables, Y)
-
+    enet = elasticNetFunc(df_variables, Y)
```
Comment (aarmey) [inline syserol/regression.py]: elasticNetFunc() isn't really a simplification anymore. Just copy the relevant lines into this function.

**Review by aarmey — CHANGES_REQUESTED** [review]
Please do this various cleanup, then have Scott review, then I'll take a look.

File: syserol/regression.py (hunk 4 above)
Comment (scottdtaylor95) [reply syserol/regression.py:54]: All is currently used in fig S1 - is that still necessary?

File: syserol/regression.py (hunk 3 above)
Comment (scottdtaylor95) [reply syserol/regression.py]: In line 60 the model is fit using ElasticNetFunc() and stored in enet, I believe

Comment (murphymadeleine21) [reply syserol/regression.py]: Yes, we fit in elasticNetFunc

**Review by aarmey — APPROVED** [review]
Simplified the code a bit. It looks like this didn't really change the predictions, but I still like the adjustments here. We can merge this, then it'd be easy to change if we need to.

**Comment by murphymadeleine21** [comment]
Okay great!

**Review by scottdtaylor95 — APPROVED** [review]

**Pull request merged**